insns.def

```diff
@@ -983,7 +983,7 @@ opt_str_uminus
 
 DEFINE_INSN
 opt_newarray_send
-(rb_num_t num, ID method)
+(rb_num_t num, ID method, VALUE arg)
```

Comment:
My original implementation instead setup a new id for `pack__buffer` and used that instead of `pack`.
I thought the double underscore would avoid ever confusing it with an actual method name.

While it was a little simpler, I wondered if maybe adding `idPack__buffer` was too weird so I changed this instruction to take an additional argument.

It could be useful in the future if we found another case to support.
I do see that `min` and `max` take an arg, so if we wanted to optimize that case we could use this argument to do that.
Until then `min`, `max`, `hash` (and the single-arg case of `pack`) are just sending a `Qnil` in its place.

I'm not sure if "separate id" had any advantages over "additional arg" so I thought I'd mention it.

Reply:
From the benchmarks I've tried performance seems the same between the two solutions.

Reply:
Also open to other suggestions for the type of the third arg.
The `VALUE` is versatile and the symbol is clear but we could use an int as a enum and avoid all the conversions 🤷

Reply:
If I understand correctly, `rb_id2sym` seems like an unneeded overhead. Why not pass `ID` for now, and consider converting it to VALUE when it really needs to be?

Reply:
Thanks, I'll do that.  As it is it looks like the other version might be a little faster so maybe that will help bring them in line.

Reply:
Now I realize that the `operands` array is typed `VALUE`
https://github.com/ruby/ruby/blob/f2728c3393d7571f681accd9c887dcd5bad65950/compile.c#L95

Trying to change it to just the id is giving me
`-e: wrong argument type Integer (expected Symbol) (TypeError)`

Reply:
If I can't change the assignment to the operands it looks like I might be able to change the type in the insn def and then drop the rb_id2sym/sym2id calls when doing the comparisons.
I'll try that and see if it measures any differently.

Reply:
We place insns and operands every 8 bytes and access it using a `VALUE *` pointer. But `ID` doesn't take more than 8 bytes, and remember, the previous operand `ID method` is already an `ID`. I don't know where the `TypeError` comes from, but it might not necessarily be a blocker.

Reply:
Interesting, thanks!

Yeah, the `ID method` is what I was referring to... I realized that was there, so I could change the type of this new arg at that same level.
The `TypeError` seems to be coming from that conversion... the `operand` is a `VALUE` and if I set the arg type on the insn to anything else I get the TypeError when the arg doesn't convert.

If I make it `ID` it does simplify the receiving code and it does work, but it then requires me to pass a symbol in the other two code paths ("min/max/hash" and "pack without buffer").

I had `Qnil` and that produces
`wrong argument type nil (expected Symbol) (TypeError)`.
If I remove the assignment to the last element I get (as you can probably guess):
`wrong argument type false (expected Symbol) (TypeError)`.

So in order to make it an `ID` I have to make up a symbol to pass to the other ones (or duplicate the `mid` arg).
This seems to be true for any type more specific than `VALUE` (since 0 works correctly for that and probably nothing else).

One thing we could do to simplify the conditions it just treat the VALUE as a boolean:
For the other code paths the arg is false, and if it's truthy we consider the buffer to be passed.
We don't need to do the symbol/ID conversion and comparison if there aren't any other possible values.

It's a little simpler, but still not great.
It makes me like the original implementation (of just using an alternate symbol) a bit more 🤔 .

Reply:
The indirection through a Symbol before it gets turned into an ID in `encoded_iseq` could be avoided if we have the instruction take a `rb_num_t` instead. The `method` operand is already acting like an enum, using ID as the underlying representation.
Changing away from ID also gets rid of the weirdness of having to make a bespoke ID with an entry in the ID table with the string content left consumed nowhere.

Reply:
Thanks!  I had considered `rb_num_t` because I thought it would be simple to just use it like an enum and the other code paths could assume 0... but I discovered they cannot, since they need to be a VALUE they have to set `operands[2] = FIX2INT(0)` or something similar, so it's not quite as clean as I was hoping.

If `rb_num_t` seems the clearest or most consistent I'm happy to switch to that.

Reply:
I'm suggesting we change the `method` operand to `rb_num_t` and add no new operand. That way you don't need to set a dummy value for the other paths as they'd each be a variation of the enum.

Reply:
Oh, ok.  Sure, I'll try that.
The code was simpler when I was just adding another case.  If we can do it that way without the extra id that sounds good to me.